modules/navigation/godot_navigation_server.cpp

```diff
@@ -676,18 +676,24 @@ bool GodotNavigationServer::agent_is_map_changed(RID p_agent) const {
 	return agent->is_map_changed();
 }
 
-COMMAND_4(agent_set_callback, RID, p_agent, Object *, p_receiver, StringName, p_method, Variant, p_udata) {
+COMMAND_4(agent_set_callback, RID, p_agent, ObjectID, p_object_id, StringName, p_method, Variant, p_udata) {
 	RvoAgent *agent = agent_owner.get_or_null(p_agent);
 	ERR_FAIL_COND(agent == nullptr);
 
-	agent->set_callback(p_receiver == nullptr ? ObjectID() : p_receiver->get_instance_id(), p_method, p_udata);
+	Object *callback_object = ObjectDB::get_instance(p_object_id);
 
-	if (agent->get_map()) {
-		if (p_receiver == nullptr) {
+	if (!callback_object) {
+		agent->set_callback(ObjectID(), p_method, p_udata);
+		if (agent->get_map()) {
 			agent->get_map()->remove_agent_as_controlled(agent);
-		} else {
-			agent->get_map()->set_agent_as_controlled(agent);
 		}
+		return;
+	}
+
+	agent->set_callback(p_object_id, p_method, p_udata);
+
+	if (agent->get_map()) {
+		agent->get_map()->set_agent_as_controlled(agent);
 	}
```

Comment:
Nitpicks:

I think the old structure can be reused. The early return makes it harder to figure out what this code block does.

There is also a validation of the object ID before passing it to `set_callback()`. I think it's redundant since `RvoAgent` still needs to validate the object ID before using it. The validation also makes the behavior inconsistent under certain circumstances (i.e. when the callback object becomes invalid before `set_callback()`, the agent is not registered as controlled in navmap; when the callback object becomes invalid after `set_callback()`, the agent won't un-register itself as controlled in navmap). But it's okay if you think this is an optimization :)

Reply:
returned old structure